src/cryptonote_core/blockchain.cpp

```diff
@@ -3666,6 +3666,19 @@ bool Blockchain::prepare_handle_incoming_blocks(const std::list<block_complete_e
   MTRACE("Blockchain::" << __func__);
   TIME_MEASURE_START(prepare);
   bool stop_batch;
+  uint64_t bytes = 0;
+
+  if(blocks_entry.size() == 0)
+    return false;
```

Comment:
That seems to return before taking the tx_lock, which cleanup_handle_x_x will unlock (regardless of what prepare_x_x_x returns).